File: src/features/transfer/csv.ts
/**
 * CSV, RFC 4180, both directions.
 *
 * The writer is `export/format.ts`'s old private `csvField` promoted, unchanged in behaviour: a
 * field is quoted when it carries a comma, a quote or a newline and **never otherwise**, so
 * `Lightning Bolt` stays `Lightning Bolt` rather than becoming `"Lightning Bolt"` on every row.
 *
 * The reader is new, and it is what makes a collection CSV a restore rather than a dump. A
 * character-by-character scanner rather than a split on commas: a quoted field may contain
 * commas *and newlines*, so there is no line-oriented shortcut that is correct.
 */

/** A field, quoted only when it has to be. An inner quote doubles — RFC 4180's escape. */
export function csvField(value: string): string {
  if (/[",\n\r]/.test(value)) return `"${value.replace(/"/g, '""')}"`;
  return value;
}

export function csvRow(values: readonly string[]): string {
  return values.map(csvField).join(",");
}

/**
 * Text into a grid of rows and fields.
 *
 * A trailing newline produces no final empty row — every file this app writes ends in one, and
 * a phantom row of blanks would become a nameless import issue for every export ever opened.
 */
export function parseCsv(text: string): string[][] {
  const rows: string[][] = [];
  let row: string[] = [];
  let field = "";
  let quoted = false;
  let i = 0;

  const endField = () => {
    row.push(field);
    field = "";
  };
  const endRow = () => {
    endField();
    rows.push(row);
    row = [];
  };

  while (i < text.length) {
    const ch = text[i];
    if (quoted) {
      if (ch === '"') {
        // A doubled quote is one quote; a lone one closes the field.
        if (text[i + 1] === '"') {
          field += '"';
          i += 2;
          continue;
        }
        quoted = false;
        i += 1;
        continue;
      }
      field += ch;
      i += 1;
      continue;
    }
    if (ch === '"') {
      quoted = true;
      i += 1;
      continue;
    }
    if (ch === ",") {
      endField();
      i += 1;
      continue;
    }
    if (ch === "\r" || ch === "\n") {
      endRow();
      // CRLF is one break, not two.
      i += ch === "\r" && text[i + 1] === "\n" ? 2 : 1;
      continue;
    }
    field += ch;
    i += 1;
  }
  // Whatever is still in hand is the last row — unless the file ended on a break, in which
  // case there is nothing in hand and no row to add.
  if (field !== "" || row.length > 0) endRow();
  return rows;
}

File: .parity-reset-backup/src-tauri-src/transfer/csv.rs
//! CSV, RFC 4180 — the writer half only.
//!
//! A port of the writer in `src/features/transfer/csv.ts`, unchanged in behaviour. `parseCsv`
//! deliberately did **not** come with it: nothing on this side ever reads a mirror file back, so a
//! reader here would exist only to be tested.

/// A field, quoted only when it has to be. An inner quote doubles — RFC 4180's escape.
///
/// The "never otherwise" half is the point rather than an optimisation: quoting unconditionally is
/// also valid RFC 4180, and it would write every `Lightning Bolt` in the mirror as
/// `"Lightning Bolt"` — noise on every row of a tree somebody reads with `git diff`.
pub fn csv_field(value: &str) -> String {
    if value.contains([',', '"', '\n', '\r']) {
        let mut out = String::with_capacity(value.len() + 2);
        out.push('"');
        for ch in value.chars() {
            if ch == '"' {
                out.push('"');
            }
            out.push(ch);
        }
        out.push('"');
        return out;
    }
    value.to_string()
}

/// One row: every field written by the rule above, joined with commas. No line ending — whatever
/// assembles the file decides what separates its rows.
pub fn csv_row(values: &[String]) -> String {
    values
        .iter()
        .map(|v| csv_field(v))
        .collect::<Vec<_>>()
        .join(",")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_plain_value_is_not_quoted() {
        assert_eq!(csv_field("Lightning Bolt"), "Lightning Bolt");
    }

    #[test]
    fn a_comma_a_quote_a_newline_and_a_carriage_return_each_force_quoting() {
        assert_eq!(
            csv_field("Borrowing 100,000 Arrows"),
            "\"Borrowing 100,000 Arrows\""
        );
        assert_eq!(csv_field("Ach! Hans, Run!"), "\"Ach! Hans, Run!\"");
        assert_eq!(csv_field("a\nb"), "\"a\nb\"");
        assert_eq!(csv_field("a\rb"), "\"a\rb\"");
    }

    #[test]
    fn an_inner_quote_doubles() {
        assert_eq!(csv_field("say \"hi\""), "\"say \"\"hi\"\"\"");
    }

    #[test]
    fn a_row_joins_with_commas_and_quotes_only_what_needs_it() {
        let row = csv_row(&["1".into(), "Bolt".into(), "a,b".into()]);
        assert_eq!(row, "1,Bolt,\"a,b\"");
    }
}
